### DISCUSSION TITLE
Build Instructions for M1/M2 version of SDR++

### DISCUSSION THREAD

**Discussion by michaelgeorge001** [AlexandreRouma/SDRPlusPlus#994] [testing]
Please note that I do not take credit for the individual instructions. They are a combination of what I have been able to piece together from the parsing of Alexandre's workflow file, other peoples experience with installing SDR++, and the oddities of HomeBrew on Apple Silicon.

This is what works for me but your mileage may vary.

**Initial Development prep setup - one time install** 

```
#install XCode Command Line Tools
#
xcode-select --install

#Rosetta may be required for some package installs
#
sudo softwareupdate --install-rosetta

#Install HomeBrew on Monterey
#
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"

#Run these two commands in your terminal to add Homebrew to your PATH:
#
(echo; echo 'eval "$(/opt/homebrew/bin/brew shellenv)"') >> /Users/admin/.zprofile
eval "$(/opt/homebrew/bin/brew shellenv)"

#Create symlinks to the Homebrew arm64 libraries and headers
#
sudo ln -s /opt/homebrew/include /usr/local/include
sudo ln -s /opt/homebrew/lib /usr/local/lib

```

**SDRPlusPlus Workflow Setup**

```
#Update brew repositories
#
brew update

#Fix stuff
#
rm -f /usr/local/bin/2to3* /usr/local/bin/idle3* /usr/local/bin/pydoc3* /usr/local/bin/python3* /usr/local/bin/python3-config*

#Install dependencies
#
arch -arm64 brew install airspy airspyhf codec2 fftw gettext glfw hackrf libbladerf libusb pkg-config portaudio rtl-sdr soapysdr volk zstd

pip3 install mako

#Clone and install libiio
#
git clone https://github.com/analogdevicesinc/libiio
cd libiio
mkdir build
cd build
arch -arm64 cmake -DCMAKE_BUILD_TYPE=Release ..
arch -arm64 make -j$(sysctl -n hw.logicalcpu)
sudo make install && cd ../../

#Clone and install libad9361
#
git clone https://github.com/analogdevicesinc/libad9361-iio
cd libad9361-iio
mkdir build
cd build
arch -arm64 cmake -DCMAKE_BUILD_TYPE=Release ..
arch -arm64 make -j$(sysctl -n hw.logicalcpu)
sudo make install && cd ../../

#Clone and install LimeSuite
#
git clone https://github.com/myriadrf/LimeSuite
cd LimeSuite
mkdir builddir
cd builddir
arch -arm64 cmake -DCMAKE_BUILD_TYPE=Release ..
arch -arm64 make -j$(sysctl -n hw.logicalcpu)
sudo make install && cd ../../

#Clone and build SDR++ git project without LimeDR or SDRPlay support.
#SDRUno does not have -arm64 Binary available.
#LimeSDR has to have the limesdr_source/CMakeLists.txt modified for the app to compile with LimeSDR support.
#
git clone https://github.com/AlexandreRouma/SDRPlusPlus.git
cd SDRPlusPlus
mkdir build
cd build
arch -arm64 cmake .. -DOPT_BUILD_PLUTOSDR_SOURCE=ON -DOPT_BUILD_SOAPY_SOURCE=ON -DOPT_BUILD_BLADERF_SOURCE=ON -DOPT_BUILD_SDRPLAY_SOURCE=OFF -DOPT_BUILD_LIMESDR_SOURCE=OFF -DOPT_BUILD_AUDIO_SINK=OFF -DOPT_BUILD_PORTAUDIO_SINK=ON -DOPT_BUILD_NEW_PORTAUDIO_SINK=ON -DOPT_BUILD_M17_DECODER=ON -DUSE_BUNDLE_DEFAULTS=ON -DCMAKE_BUILD_TYPE=Release
arch -arm64  make -j$(sysctl -n hw.logicalcpu)

#Build the SDR++ Mac Application
cd ..
sh make_macos_bundle.sh ~/Dev/SDRPlusPlus/build ./SDR++.app
zip -r sdrpp_macos_silicon.zip SDR++.app

```

These instructions are used to create the program on a M1 Mac mini running MacOS Monterey. I can only test a few source plugins such as rtl-tcp, rtl-sdr, and sdrpp_server. I do have a SDRPlay RSPdx but can only use it with the Intel nightly build.

PS. Do not run SDR++ in a macOS Monterey or newer VM. It will compile the code and will create a properly functioning binary. However, SDR++ is indirectly dependent on OpenGL for displaying, and Apple Silicon macOS VMs support neither OpenGL nor OpenCL.

I can attach a file of the program if Alexandre approves it [I do realize that he has no way of testing or validating it.]

-Mike-

**Comment by stedrow** [comment]
Thanks for this write-up, I've updated the instructions a bit for macOS Ventura below.

```
Updated for macOS Ventura (13.3)

# Changes:
* Installing `rtaudio` and `cmake` with brew
* Modifying codec2 version.h include post brew install
* Removing `arch -arm64` prefix on commands, this isn't needed anymore since it's the default arch.

# Update brew repositories
#
brew update

# Install dependencies
#
brew install airspy airspyhf codec2 cmake fftw gettext glfw hackrf libbladerf libusb pkg-config portaudio rtaudio rtl-sdr soapysdr volk zstd

# Clone and install libiio -- Required dependency
#
git clone https://github.com/analogdevicesinc/libiio
cd libiio
mkdir build
cd build
cmake -DCMAKE_BUILD_TYPE=Release ..
make -j$(sysctl -n hw.logicalcpu)
sudo make install && cd ../../

# Clone and install libad9361 -- Required dependency
#
git clone https://github.com/analogdevicesinc/libad9361-iio
cd libad9361-iio
mkdir build
cd build
cmake -DCMAKE_BUILD_TYPE=Release ..
make -j$(sysctl -n hw.logicalcpu)
sudo make install && cd ../../

# Fix codec2 brew `version.h` issue (pops up during make command)
#
Details of issue can be found here:  
https://github.com/AlexandreRouma/SDRPlusPlus/issues/730#issuecomment-1165987441

TL;DR is that we need to replace line 32 of `/opt/homebrew/Cellar/codec2/1.0.5/include/codec2/codec2.h`
(note the 1.0.5 version may be different for you depending on what brew installed so the path may differ)

Comment out line 32 so it looks like this:
/*#include <codec2/version.h>*/

And add this line below it:
#include <version.h>

# Clone and build SDR++ git project without LimeDR or SDRPlay support.
#
git clone https://github.com/AlexandreRouma/SDRPlusPlus.git
cd SDRPlusPlus
mkdir build
cd build
cmake .. -DOPT_BUILD_PLUTOSDR_SOURCE=ON -DOPT_BUILD_SOAPY_SOURCE=ON -DOPT_BUILD_BLADERF_SOURCE=ON -DOPT_BUILD_SDRPLAY_SOURCE=OFF -DOPT_BUILD_LIMESDR_SOURCE=OFF -DOPT_BUILD_AUDIO_SINK=OFF -DOPT_BUILD_PORTAUDIO_SINK=ON -DOPT_BUILD_NEW_PORTAUDIO_SINK=ON -DOPT_BUILD_M17_DECODER=ON -DUSE_BUNDLE_DEFAULTS=ON -DCMAKE_BUILD_TYPE=Release
make -j$(sysctl -n hw.logicalcpu)

# Build the SDR++ Mac Application
cd ..
sh make_macos_bundle.sh $PWD/build ./SDR++.app
zip -r sdrpp_macos_silicon.zip SDR++.app

# Copy SDR++.app to your Applications directory
cp SDR++.app /Applications/SDR++.app
```

**Reply by michaelgeorge001** [reply]
Very good.

I actually had updated my instructions for my own use but just never got around to updating them here.

I had run into the codec2 problem as well but I just went with compiling it from git rather than homebrew.

```
git clone https://github.com/drowe67/codec2.git && cd codec2 && mkdir build && cd build
cmake .. -DCMAKE_BUILD_TYPE=Release
make -j$(sysctl -n hw.logicalcpu)
sudo make install && cd ../../
```

LimeSDR support can added if you modify the 'limesdr_source->CMakeLists.txt' file just before executing the SDR++ 'make' command.

This is what worked for me:

Replace:

```
    # Not in pkg-config
    target_link_libraries(limesdr_source PRIVATE LimeSuite)
```
With:

```
    pkg_check_modules(LIBLIMESDR REQUIRED LimeSuite)

    target_include_directories(limesdr_source PRIVATE ${LIBLIMESDR_INCLUDE_DIRS})
    target_link_directories(limesdr_source PRIVATE ${LIBLIMESDR_LIBRARY_DIRS})
    target_link_libraries(limesdr_source PRIVATE ${LIBLIMESDR_LIBRARIES})
```

-Mike-

**Reply by K0HYD** [reply]
Thanks for publishing the Ventura code. I'm running Sonoma now on an M1 MacBook Pro. I have been running the Intel code from the nightly build successfully but would like to compile a native version for the M1 if that's possible. Has this script been approved yet? How well does it run compared to the Intel nightly? Would like to experiment if it's safe but thought I would ask if there had been any changes to the script since April first. Again, Thanks for your work. Dale - K0HYD

**Reply by michaelgeorge001** [reply]
These current build instructions do **not** work, at least for me. Dependencies appear to have changed just enough that the compile fails, and I haven't had enough time in my life to isolate the specific issues and rewrite the instructions.

In my case, it always seems to be the SDRPlay and LimeSuite devices that cause compilation issues for me.

For now, I suggest using the Intel nightly build - it works.

I can't speak for Alexandre but he is probably very busy with his life just like the rest of us, and that hopefully he will release new code updates as his time allows.

Happy Holidays,
Mike

**Comment by Aricg** [comment]
Comments inline of changes I needed on Sonoma
```
git clone https://github.com/analogdevicesinc/libiio
cd libiio
#main is 'next' now and broken
git checkout libiio-v0 #maintenance branch 
mkdir build
cd build
cmake -DCMAKE_BUILD_TYPE=Release ..
make -j$(sysctl -n hw.logicalcpu)
sudo make install && cd ../../

git clone https://github.com/analogdevicesinc/libad9361-iio
cd libad9361-iio
mkdir build
cd build


#Y'all have a /usr/local/include/ on your mac?
cmake -DCMAKE_BUILD_TYPE=Release \
      -DCMAKE_C_FLAGS="-I/Library/Frameworks/iio.framework/Versions/0.26/Headers" ..

make -j$(sysctl -n hw.logicalcpu)
```

**Comment by Perterra1** [comment]
Hi, I'm not as technical as you guys yet. I have downloaded the Apple Silicon version of SDR++, and I have plugged in my new NESDR smart v5 the device does not seem to be connecting with the software. The NESDR does not require a new driver because it runs natively. What do I need to do?
Edit: thanks in advance
